docker/docker_client.go

```diff
@@ -428,15 +379,34 @@ func getAuth(ctx *types.SystemContext, registry string) (string, string, error)
 }
 
 type pingResponse struct {
-	WWWAuthenticate string
-	APIVersion      string
-	scheme          string
+	APIVersion string
+	scheme     string
+}
+
+func (c *dockerClient) saveChallenges(chs []challenge, u *url.URL) {
+	urlCopy := url.URL{
+		Path:   u.Path,
+		Host:   u.Host,
+		Scheme: u.Scheme,
+	}
+	normalizeURL(&urlCopy)
+	c.challenges[urlCopy.String()] = chs
+}
+
+func (c *dockerClient) getChallenges(req *http.Request) []challenge {
+	ping := url.URL{
+		Host:   req.URL.Host,
+		Scheme: req.URL.Scheme,
+		Path:   "/v2/",
```

Comment:
How is this, at the high level, supposed to help with #191 / #195 ? We still use the `WWW-Authenticate` value from `/v2/`, only now there is a lot more infrastructure around storing and retrieving the value.